core/execution/spot/market.go

```diff
@@ -3387,3 +3487,60 @@ func (m *Market) CheckOrderSubmissionForSpam(orderSubmission *types.OrderSubmiss
 		orderSubmission.Type,
 		quantumMultiplier)
 }
+
+func (m *Market) enterAutomatedPurchaseAuction(ctx context.Context, orderSide types.Side, orderPrice *num.Uint, orderSize uint64, reference string, duration time.Duration) (string, error) {
+	if !m.canTrade() {
+		return "", fmt.Errorf(fmt.Sprintf("cannot trade in market %s", m.mkt.ID))
+	}
+
+	// if we're in governance auction we're not placing automated purchase orders because we don't have
+	// control over when the market will be resumed
+	if m.mkt.TradingMode == types.MarketTradingModeSuspendedViaGovernance {
+		return "", fmt.Errorf(fmt.Sprintf("cannot enter auction in market %s - market is in governance suspension", m.mkt.ID))
+	}
+
+	if m.as.InAuction() {
+		now := m.timeService.GetTimeNow()
+		aRemaining := int64(m.as.ExpiresAt().Sub(now) / time.Second)
+		if aRemaining >= int64(duration.Seconds()) {
+			return "", nil
+		}
+		m.as.ExtendAuctionAutomatedPurchase(types.AuctionDuration{
+			Duration: int64(duration.Seconds()) - aRemaining,
+		})
+		if evt := m.as.AuctionExtended(ctx, now); evt != nil {
+			m.broker.Send(evt)
+		}
+	} else {
+		m.as.StartAutomatedPurchaseAuction(m.timeService.GetTimeNow(), int64(duration.Seconds()))
+		m.mkt.TradingMode = types.MarketTradingModeAutomatedPuchaseAuction
+		m.mkt.State = types.MarketStateSuspended
+		m.enterAuction(ctx)
+		m.broker.Send(events.NewMarketUpdatedEvent(ctx, *m.mkt))
+	}
+	// if we were able to enter an auction, now place the order for the network account
+	if m.as.InAuction() {
+		os := &types.OrderSubmission{
+			MarketID:    m.mkt.ID,
+			Price:       orderPrice.Clone(),
+			Size:        orderSize,
+			Side:        orderSide,
+			TimeInForce: types.OrderTimeInForceGFA,
+			Type:        types.OrderTypeLimit,
+			Reference:   reference,
+		}
+		conf, err := m.SubmitOrder(ctx, os, types.NetworkParty, crypto.RandomHash())
```

Comment:
Is `crypto.RandomHash()` here not going to cause snapshot issues, or events being sent to the data-node being different?

Reply:
why snapshot issues - the order id is persistent
why being sent to different data nodes?

Reply:
`crypto.RandomHash()` will end up being used to generate the order ID no? Which will be different across different nodes. It'll get saved into a snapshot if one happens during the auction period, and also any event the data-node sends out for that order will have the random ID in it.

Reply:
I forgot the random hash doesn't take have a seed, will fix it.